auth0/src/main/java/com/auth0/android/provider/OAuthManager.java

```diff
@@ -0,0 +1,295 @@
+        }
+    }
+
+    public boolean resumeAuthorization(AuthorizeResult data) {
+        if (!data.isValid(requestCode)) {
+            Log.w(TAG, "The Authorize Result is invalid.");
+            return false;
+        }
+
+        final Map<String, String> values = CallbackHelper.getValuesFromUri(data.getIntent().getData());
+        if (values.isEmpty()) {
+            Log.w(TAG, "The response didn't contain any of these values: code, state, id_token, access_token, token_type, refresh_token");
+            return false;
+        }
+        logDebug("The parsed CallbackURI contains the following values: " + values);
+
+        try {
+            checkErrorValue(values.get(KEY_ERROR));
+            checkStateValue(values.get(KEY_STATE));
+            checkNonceValue(values.get(KEY_NONCE), values.get(KEY_ID_TOKEN));
+        } catch (AuthenticationException e) {
+            callback.onFailure(e);
+            return true;
+        }
+
+        Log.d(TAG, "Authenticated using web flow");
+        final Credentials urlCredentials = new Credentials(values.get(KEY_ID_TOKEN), values.get(KEY_ACCESS_TOKEN), values.get(KEY_TOKEN_TYPE), values.get(KEY_REFRESH_TOKEN));
+        if (!shouldUsePKCE()) {
+            callback.onSuccess(urlCredentials);
+            return true;
+        }
+
+        //Finish Code Exchange
+        pkce.getToken(values.get(KEY_CODE), new AuthCallback() {
+            @Override
+            public void onFailure(@NonNull Dialog dialog) {
+                callback.onFailure(dialog);
+            }
+
+            @Override
+            public void onFailure(AuthenticationException exception) {
+                callback.onFailure(exception);
+            }
+
+            @Override
+            public void onSuccess(@NonNull Credentials codeCredentials) {
+                callback.onSuccess(mergeCredentials(urlCredentials, codeCredentials));
+            }
+        });
+        return true;
+    }
+
+    //Helper Methods
+
+    private void checkErrorValue(String errorValue) throws AuthenticationException {
+        if (errorValue == null) {
+            return;
+        }
+        Log.e(TAG, "Error, access denied. Check that the required Permissions are granted and that the Application has this Connection configured in Auth0 Dashboard.");
+        if (ERROR_VALUE_ACCESS_DENIED.equalsIgnoreCase(errorValue)) {
+            throw new AuthenticationException(ERROR_VALUE_ACCESS_DENIED, "Permissions were not granted. Try again.");
+        } else {
+            throw new AuthenticationException("a0.invalid_configuration", "The application isn't configured properly for the social connection. Please check your Auth0's application configuration");
+        }
+    }
+
+    private void checkStateValue(String receivedState) throws AuthenticationException {
+        String sentState = parameters.get(KEY_STATE);
+        if (receivedState != null && sentState != null && !receivedState.equals(sentState)) {
```

Comment:
We should only deem the response `state` as valid if it's value matches the one in the request, either if both are `null` or the same `String`